### ISSUE TITLE
[HOLD for payment 2024-03-22] [$1000] Settings - Performance and heat issues on the "Troubleshoot" option

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#36645]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
Issue was found when executing #35306
**Version Number:** 1.4.42.0
**Reproducible in staging?:** y
**Reproducible in production?:** New feature
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause internal team
**Slack conversation:**

## Action Performed:
1. Navigate to Account settings - About - Troubleshoot
2. Background the app
3. Foreground the app
4. Swipe up and down

## Expected Result:
Swiping should be smooth. The phone shouldn't heat up or crash.

## Actual Result:
Low performance when I foreground the app. The phone heats up if I stay on the page. It even crashed once.
The issue disappears if I navigate back and to the "Troubleshoot" again.
## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [x] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/00fb81f1-f0b8-4114-890f-a6ca371e338d

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01dd3351f4b07d85c0</li>
        <li>Upwork Job ID: 1758255599540744192</li>
        <li>Last Price Increase: 2024-02-27</li>
    </ul>
</details>

**Comment by marcaaron** [comment]
Trying to repro now.

**Comment by marcaaron** [comment]
It seems like a possible memory leak or something. I do notice things get slightly more janky. But doesn't seem like a blocker if the only way to reproduce it is to go to this particular page and background the app.

**Comment by laurenreidexpensify** [comment]
Asking [here](https://expensify.slack.com/archives/C01SKUP7QR0/p1708357757696129) to get another repro of this

**Comment by mallenexpensify** [comment]
@laurenreidexpensify , snaggin' from you since I have an iPhone and am able to reproduce, at least the part where the scroll isn't smooth. 

https://github.com/Expensify/App/assets/22508304/62851b28-8abf-4f93-93fe-848af93c40ba

**Comment by mallenexpensify** [comment]
Also... reporting back that I didn't experience any heat issues or noticeable battery drain.

**Comment by mallenexpensify** [comment]
Waiting for proposals

**Comment by mallenexpensify** [comment]
Thanks for posting in #expensify-open-source @hoangzinh 
Doubling price to get eyes and proposals.

**Comment by wustwg** [comment]
Does only iOS have this problem？Does Android work smoothly?

**Comment by wustwg** [comment]
Contributor details
Your Expensify account email: wg940412478@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01eec1faf4fd4ceb3c?viewMode=1

**Comment by tamachai** [comment]
Contributor details
Your Expensify account email: poppies.great_0r@icloud.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0107f7a8dc488440a0

**Comment by tamachai** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Performance degradation occurs when viewing the Troubleshoot screen on iOS.

### What is the root cause of that problem?

The Troubleshoot screen is presented using `transparentModal`. This causes the About screen's animation to continue to play as the Troubleshoot screen overlays the About screen.

### What changes do you think we should make in order to solve the problem?

Update the presentation type for the Troubleshoot screen to `modal`. This will prevent rendering the About screen while the Troubleshoot screen is displayed.

### What alternative solutions did you explore? (Optional)

- Tested playing/pausing animations based on `useFocusEffect`, but this introduces stutter when an animation is resumed.

- It might be worth optimizing the `Coin.lottie` (569KB) and `Desk.lottie` (165KB) animations. Apart from `ExpensifyLounge.lottie` (313KB), all other animations are <100KB.

**Comment by ikevin127** [comment]
⚙ From digging into it, what I can say so far is that this issue is **not** caused by:
- the Lottie animation
- the `ReconnectApp` call upon getting the app back to the foreground
- the `triggerReconnectionCallbacks` loop

📜 I can say this w/ certainty because I commented out all of the above logic and the issue is still there.

👁 **Observations**:
- I tested locally on a real device (iPhone 12mini) and it does overheat when getting the app back to the foreground and leaving it there for ~ 5mins
- weirdly, this seems to only be happening on the Troubleshoot page (to this extent) when it comes to both laggy UI and overheating -> doing the same thing on say About page, we still get some minor UI lagging but no overheating
- when on the Troubleshoot page after getting the app back to the foreground, if navigating back to the About page, the issue is gone, it's like engaging w/ the navigation stack makes the issue go away instantly

I'm still digging and will come up with a proposal if I figure out the root cause and can come up w/ a decent solution.

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

Performance and heating issues on the Troubleshoot page if we put the app into background state then back to foreground.

### What is the root cause of that problem?

> [!NOTE]
> The issue is only happening on native devices because the navigation is stacking the screens (in memory).
> I wasn't able to reproduce on Android, only on iOS.

This happens because on Native when we navigate to a new screen, the previous screen remains in the navigation stack. Each new screen is added (pushed) onto the top of the stack, and when the user navigates back, the current screen is removed (popped), revealing the previous screen.

The problem in our case is that some of these pages contain Lottie animations and because the screens are kept in memory as we navigate (ex. from About to Troubleshoot), both pages animations never unmount when transitioning the app back from background to foreground - which increases the load on the UI thread significantly.

### What changes do you think we should make in order to solve the problem?

We can handle this by creating a **native platform specific** hook called `useAppState` which we will use within the [`Lottie/index.tsx`](https://github.com/Expensify/App/blob/04c830536bd98b4f682698f92ee0782900bfdeb8/src/components/Lottie/index.tsx#L14) component.

<details>
<summary>useAppState hook code</summary>

```ts
// src/hooks/useAppState/index.native.ts

import React from 'react';
import type { AppStateStatus } from 'react-native';
import { AppState } from 'react-native';

type AppStateType = {
    isForeground: boolean;
    isInactive: boolean;
    isBackground: boolean;
}

function useAppState() {
    const [appState, setAppState] = React.useState<AppStateType>({
        isForeground: AppState.currentState === 'active',
        isInactive: AppState.currentState === 'inactive',
        isBackground: AppState.currentState === 'background',
    });

    React.useEffect(() => {
        function handleAppStateChange(nextAppState: AppStateStatus) {
            setAppState({
                isForeground: nextAppState === 'active',
                isInactive: nextAppState === 'inactive',
                isBackground: nextAppState === 'background',
            });
        }
        const subscription = AppState.addEventListener('change', handleAppStateChange);

        return () => subscription.remove()
    }, []);

    return appState;
}

export default useAppState;

```

</details>

Which we will add to the Lottie component like so:

```tsx
// ...existing logic
const appState = useAppState();

// If the image fails to load or app is in background state, we'll just render an empty view
// using the fallback in case of a Lottie error or to prevent memory leaks.
if (isError || appState.isBackground) {
    return <View style={aspectRatioStyle} />;
}
// ...existing logic
```

By doing this we unmount the animations when the app goes into background, and mount them again when the app is foregrounded - this way we take the load off the UI thread and get rid of both performance and heating issues.

### Videos

<details>
<summary>iOS: Native</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/13825eeb-3a70-4629-b74c-fa1da3a9ce87"> | <video src="https://github.com/Expensify/App/assets/56457735/e2edbbe4-45f0-4983-9e49-70c6526ddfbb"> |

</details>

**Comment by ikevin127** [comment]
@hoangzinh With the solution I proposed that is the trade-off since we're re-mounting the animation.

This load delay also happens first time you open the app and navigate to one of the screens where an animation exists - so it makes sense it would also exist when bringing the app back from background - which means all processes restart / re-mount.

I can look into more but for now just assume this is as good as it gets considering the RCA.

**Comment by tamachai** [comment]
> @tamachai @ikevin127 Thanks for proposals everyone. Both proposals look great, especially @ikevin127 with details root cause. I have a few concerns about solutions:
> 
> * @tamachai: `Update the presentation type for the Troubleshoot screen to modal` I'm worried about this approach potentially causing regression bugs. We must have reasons of using `transparentModal`. There is a PR change to use `card` presentation but has been reverted [revert: native stack #37305](https://github.com/Expensify/App/pull/37305).
>   About your alternative solutions, could you elaborate more about `playing/pausing animations based on useFocusEffect`? Thanks
> * @ikevin127 When I watch your recordings, when the app is foregrounded, the animation doesn't load in a short period. Is there any way we can avoid it? Thanks

`card` uses OS defined animations. Whereas `modal` and `transparentModel` use the same animations as each other but have distinct settings to control the visibility of the header and previous screen. It's best to avoid transparent modal unless the app needs to render the screen below the current, such as when overlaying modals on iPad or using effects like blur. For reference, here's the view hierarchy with `transparentModal` vs `modal`.

**transparentModal:**
<img width="1070" alt="transparentModal" src="https://github.com/Expensify/App/assets/160195832/4d845141-7d7b-4aae-abfc-0b406ccb091b">

**modal:**
<img width="970" alt="modal" src="https://github.com/Expensify/App/assets/160195832/d05536f7-8c30-48af-a12a-e962c99727ff">

When I experimented playing/pausing the animation via `useFocusEffect`, a stutter was introduced during view transition, as the animation begins to play.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by mallenexpensify** [comment]
Reproducible for me on 1.4.47


https://github.com/Expensify/App/assets/22508304/77e8a3a8-163c-4cac-ae5b-e708562017d4

**Comment by tamachai** [comment]
> @tamachai if we use `modal`, will it keep the previous state if we navigate back? For example, previously we had an expectation here #36077

Yes, it should behave the same way. It doesn’t actually alter the view stack. An issue would occur if the modal wasn’t full screen, but this isn’t the case. Maybe we can look at only applying this to the settings and troubleshooting screens and then verify there are no regressions.

> > playing/pausing the animation via `useFocusEffect`
> 
> Could you elaborate on how would you `playing/pausing`?

LottieView exposes a ref prop that can be used to control the view via its imperative API. Using useFocusEffect we can call play/pause, depending on if the screen hosting the Lottie view is focused. The issue here is that starting playback introduces stutter. This stutter is especially noticeable when swiping back to the previous screen.

**Comment by ikevin127** [comment]
@hoangzinh I tried that but for my proposal it didn't work since (IMO) our issue has to be tackled based on the AppState change (background/foreground) and not whether the screen with animation is focused or not.

As explained by the other proposal's author, because the way our navigation works currently - there's no page focus change when the app goes into background and comes back to the foreground, hence we cannot use that to tackle the issue as you suggested.

**Comment by tamachai** [comment]
> @ikevin127 @tamachai could you guys try to use `useFocusEffect` to hide/show lotte view and share recordings results here? So we can know which approach gives better results. Thanks

Sure thing. I’ll try to put something together today.

**Comment by mallenexpensify** [comment]
Has anyone confirmed if there is battery drain or if performance is affected. Wonder if this purely aesthetic. Asking because I'm wanting to assign this too a VIP/Wave project.

**Comment by ikevin127** [comment]
I can confirm that when the UI thread gets really heavy upon resuming app from background to foreground -> if waiting for 3+ minutes the phone heat increases gradually w/ every minute which also drains battery faster than usual.

**Comment by mallenexpensify** [comment]
Thanks @ikevin127 , added to #vip-vsb

**Comment by ikevin127** [comment]
@hoangzinh It does fix the issue - both heating and heavy animation problems were gone when I tested my solution.

Again, as explained in [my proposal](https://github.com/Expensify/App/issues/36645#issuecomment-1971281335)'s RCA - the lottie animations never stop when the app goes into the background, causing memory leak - the issue occurs because when the app reopens (from background state), we are presented with all that in-memory build-up - which causes our heavy UI / overheating issues.

What my solution does is un-mount the animation as soon as the app goes into the background and re-mounts it when the app is brought back to the foreground.

The only tradeoff of the solution being that there's a bit of delay until animation is rendered, depending on animation size - which we already have currently when first navigating to any screen with Lottie animation.

**Comment by hayata-suenaga** [comment]
Thank you all for your help with this bug. I have a follow-up question regarding @ikevin127's analysis of the issue.

> The Lottie animation doesn't unmount or stop when the app moves to the background or when the screen (i.e., the stack) is hidden behind another during navigation.

I have two questions:

- Are we certain that the Lottie library doesn't manage the animation state when the app moves to the background?
- We use `react-freeze`. Do we know why this isn't resolving the performance issue when hidden stack screens have Lottie animations?

**Comment by ikevin127** [comment]
@hayata-suenaga Here's my take:

> Are we certain that the Lottie library doesn't manage the animation state when the app moves to the background?

**TLDR**: The problem lies in the way our screens are stacked when transitioning the app to background state and back, not an issue with the Lottie library.

Lottie does manage animation in terms of app state by keeping record of the animation timeline and resuming, according to [this](https://github.com/lottie-react-native/lottie-react-native/pull/980) `lottie-react-native` merged PR - but other than that there's no un-mount / re-mount functionality for edge cases like ours (navigation stack `transparentModal` + app state change with Lottie animation ongoing).

> We use react-freeze. Do we know why this isn't resolving the performance issue when hidden stack screens have Lottie animations?

As mentioned in our docs [here](https://github.com/Expensify/App/blob/70bfa9452413f6ac542d2482e78476891674a7a7/contributingGuides/NAVIGATION.md?plain=1#L76):

> `react-freeze` allows us to increase performance by avoiding unnecessary re-renders of screens that aren’t visible to the user anyway.

This only prevents re-renders but since in our edge case there's no re-rendering happening, but only animation related behaviour (UI thread) - which does not help when bringing app back from background to the foreground and processes resume.

We need the proposed solution which un-mounts and re-mounts the animation component in order to actually fix this performance issue.

**Comment by hayata-suenaga** [comment]
sorry I always click a wrong button 😓

**Comment by ikevin127** [comment]
PR ready for review! 🚀

**Comment by mallenexpensify** [comment]
@ikevin127 , @hoangzinh , hired ya both, can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~01dd3351f4b07d85c0

**Comment by ikevin127** [comment]
Accepted, thank you!

**Comment by ikevin127** [comment]
@mallenexpensify All good for payment here ?

**Comment by mallenexpensify** [comment]
Contributor: @ikevin127 paid $1000 via Upwork
Contributor+: @hoangzinh paid $1000 via Upwork. 

@hoangzinh plz fill out the BZ checklist above. 

Thx!

**Comment by mallenexpensify** [comment]
Thanks @hoangzinh
